src/tests/end2end/D3D12ResidencyManagerTests.cpp

```diff
@@ -300,14 +301,22 @@ TEST_F(D3D12ResidencyManagerTests, OverBudgetUsingBudgetNotifications) {
     ASSERT_SUCCEEDED(ResourceAllocator::CreateAllocator(
         CreateBasicAllocatorDesc(), residencyManager.Get(), &resourceAllocator));
 
-    const D3D12_RESOURCE_DESC bufferDesc = CreateBasicBufferDesc(GPGMM_MB_TO_BYTES(1));
+    constexpr uint64_t kBufferMemorySize = GPGMM_MB_TO_BYTES(1);
+    const D3D12_RESOURCE_DESC bufferDesc = CreateBasicBufferDesc(kBufferMemorySize);
+
+    ALLOCATION_DESC bufferAllocationDesc = {};
+    bufferAllocationDesc.HeapType = D3D12_HEAP_TYPE_DEFAULT;
+
+    const DXGI_MEMORY_SEGMENT_GROUP bufferMemorySegment =
+        residencyManager->GetMemorySegmentGroup(bufferAllocationDesc.HeapType);
+    const uint64_t memoryUnderBudget = GetBudgetLeft(residencyManager.Get(), bufferMemorySegment);
 
     // Keep allocating until we reach the budget.
     std::vector<ComPtr<ResourceAllocation>> allocations = {};
-    while (!IsOverBudget(residencyManager.Get())) {
+    while (resourceAllocator->GetInfo().UsedMemoryUsage + kBufferMemorySize < memoryUnderBudget) {
```

Comment:
I'm seeing flaky failures with this test. I put a breakpoint on the budget update notification callback and I'm seeing the budget fluctuate as this loop runs. You won't be able to guarantee all allocations are resident (for when you test later) if you aren't dynamically updating the budget in your loop condition.

Reply:
Ack, fixed. Can you please verify?

Reply:
Weird. If you didn't notice, the bots just flaked out on me.

https://github.com/intel/GPGMM/runs/7363823311?check_suite_focus=true

Reply:
Weird. If we were over budget, then we shouldn't have entered the loop and given the manager an opportunity to evict. Unless the budget change could have happened in that tiny amount of time between checking the loop condition and doing the allocation.

Also - after looking more I'm not sure the while condition makes sense in the latest commit. Every loop you call GetBudgetLeft, which is the amount of memory we have remaining to allocate and its compared with the total memory size after the next allocation, which I don't think makes sense. Shouldn't it be `while(GetBudgetLeft() - kBufferMemorySize > 0)` ?